File: graycode.java
public class Solution {
    public ArrayList<Integer> grayCode(int n) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if(n==0) {
            ArrayList<Integer> res = new ArrayList<Integer>();
            res.add(0);
            return res;
        }
        
        ArrayList<Integer> a = grayCode(n-1);
        int s = 1<<(n-1);
        
        for(int i=a.size()-1;i>=0;i--){
            a.add(a.get(i)|s);    
        }
        return a;
    }
}

File: cpp/GrayCode.cpp
class Solution {
public:
    vector<int> grayCode(int n) {
        if(n==0) {
            return vector<int>{0};
        }
        if(n==1) {
            return vector<int> {0,1};
        }
        vector<int> v = grayCode(n-1);
        vector<int> v1(v.begin(), v.end());
        reverse(v1.begin(), v1.end());
        int c = (1 << (n-1));
        for(int i=0;i<v1.size();i++) {
            v1[i] = v1[i] + c;
        }
        v.insert(v.end(), v1.begin(), v1.end());
        return v;
    }
};
